Add dance.selections.allowEmpty configuration.

yuchenshi opened the pull request:
This defaults to true (existing behavior). When set to false, all selections keep at least one character.

Fixes #44.

Note that even with this change, the selection is not fully Kak/Vim-like. In kak, one-character selections have the same anchor and cursor and thus can be extended to either side, where as in VSCode one-character selections can be either forward or backwards, and will respond differently to expansions in Dance. Fixing that will require a much larger PR.

71 commented on src/commands/index.ts:
```diff
@@ -157,9 +157,15 @@ export class CommandDescriptor<Input extends InputKind = InputKind> {
     }
 
     if (flags & CommandFlags.ChangeSelections) {
+      state.normalizeSelections(editor)
```
This should be in the `onDidChangeTextEditorSelection` event handler in `extension.ts` (for instance, when the user clicks or the selection changes for another reason).

yuchenshi replied:
Thanks for the suggestion. This was also my first thought but it turned out to be quite laggy and can race with itself, especially when making a selection with mouse dragging. I believe this is due to `onDidChangeTextEditorSelection` triggering too frequently when selections are rapidly changing. Also, it introduces an annoying flicker with selections whenever you do anything (h, j, k, l, etc.). Feel free to try it out at  https://github.com/yuchenshi/dance/tree/keep-selection-non-empty.

The current way of normalizing selection post-command is the most reasonable trade-off that I can think of and I'd suggest we go with this. I can address the other comments now that I'm back to town.

71 replied:
I'd be interested on getting back more feedback about this, because it works flawlessly on my end with those changes. Mouse selections, regular selections, etc. The only "annoying thing" is that it highlights all characters matching the one under the cursor, but this can likely be disabled when we force the 1-character selection, and re-enabled whenever a bigger selection is made by the user.

yuchenshi replied:
Sure, let me implement all the performance improvements you've suggested and then try the `onDidChangeTextEditorSelection` approach again to see if it gets better. I think the flickering may be correlated with allocations & GC so we may not even need debouncing if it gets fast enough.

71 replied:
FYI, I edited my comment above. Thanks again for your work!

Does your computer have good specs? I don't have any delays on my end but I have quite a good setup.

yuchenshi replied:
I got rid of unnecessary allocations and assignments and I think dragging related issues are gone at least on my end. (My computer doesn't have good specs.) I assume we're safe without the timeout but feel free to test it out on your computer.

There's still a minor annoyance that when pressing `h`, `l`, the current line background flickers for a split second, since technically selection changed to empty and then 1-character, and the background is a little bit different in each case. I can live with it though.

yuchenshi replied:
Okay, so after using it for a while, more issues came up. Example: `o` will now select a new character on a new line in addition to entering insert mode, since the selection change triggered in between. This gets really annoying and unpredictable. @71 thoughts?

71 replied:
To me, changing selections post-command won't cut it. Either we manage to get this working, or we find another way (e.g. have different hard-coded behavior when the selection is empty, compatible with block cursor emulating a one-char selection).

Sorry for the delay, I'll try your new changes.

71 replied:
1. That flicker should be easy to remove if the theme is changed to make the background the same for empty and non-empty selections.
2. I added a new `ignoreSelectionChanges` boolean to `Extension` that I set to `true` when executing a command, and back to `false` afterwards. After setting it back to false, I re-execute `normalizeSelections`. This fixes the issues regarding `o` and `c` on my end.

71 replied:
See branch https://github.com/71/dance/tree/allow-empty-selections for my own changes.

yuchenshi replied:
Thanks for the feedback. I think I'm convinced about the points about `onDidChangeTextEditorSelection` and let's keep it that way. I've merged your changes into my branch.

I'm far less convinced on [forcing to select the last character of line](https://github.com/71/dance/commit/211fe4d73d9bf3d1045d5a95ce0444930dd6c7fd#diff-45327f86d4438556066de133327f4ca2R302-R307) though. First, the new behavior is rather Vim-like than Kakoune-like, and removes some of the power of moving multiple cursors across lines to make changes. In simple cases, I found it much harder to select and delete the line break. It's also really annoying if I press `h` on the beginning of the line, I cannot simply go back by pressing `l`. I'd say we revert this but keep the special casing you did for deleting empty lines.

On a side note, I think we're almost there -- in the last few weeks we've identified and addressed most of the problems and we're reaching towards the local optimum here. I don't want to be bikeshed too much on the exact semantics here but I do hope that we address the point in the last paragraph. I'll then move on to finalize the PR and fix any issues that other testers may find soon. Just let me know if you need more help on anything!

yuchenshi replied:
Update: I've created a new commit 039eaa8752e2e8c79ae664fd4e7a67694aaad7bc to demonstrate another idea for handling end of line situations. In short, this won't select the line break (so cursor is still at this line), but pressing `c`/`d` deletes the line break as you'd expect. It also covers the empty line case.

71 replied:
I don't understand your problem regarding `h` and `l`. Can you show me a specific sequence of keys that doesn't work out well? I've tried changing lines using `h` and `l` and didn't have any issue.

I played a little bit with your commit and didn't notice many changes from mine, though.

I just found two other issues:
1. Selecting with the mouse doesn't work in most cases, I think you'll have to add the timeout as shown earlier (e.g. try selecting a character and the going down with your mouse pointer -- the first character you selected will be deselected).
2. Pressing `s-l` does not work in one-character selections because it deselects the one character, which leads to the selection being empty and to Dance re-selecting the same character; therefore you can't extend selections to the right.

FYI, before you submit your code I'll have you change everything to respect the overall coding style (e.g. remove semicolons, add a little more whitespace).

yuchenshi replied:
Sure, let me answer the first question regarding `h` and `l` first, and I'll get to the mouse selection fix and code styles in the mean time.

Say we have text like this:
```
abcdef
123
```

First, place the cursor at `1` (selecting `1`). Now press `h`. The cursor goes to `f` on the previous line. At this point, pressing `l` won't do anything, where I'd expect it to undo what I just did with `h` and move to the next line (to select `1` again).

Also, as a kak user I would want to be able to select the line break by pressing `h` / `l`, but I understand that in VSCode such a selection would cross two lines, creating quite a few issues. My latest commit addresses that as explained above.

yuchenshi replied:
I've added the mouse timeout and it's working fine for me. I think the 200ms timeout is reasonable and let's stick with that. I've also improved the code style and please feel free to take a look and drop some comments.

For the second point, I've mentioned that in the PR description that a fully Kak-like fix will be to treat one-character selection as non-directional. In terms of implementation, Dance will need to treat such selections as forward when extending forward, which is a much larger PR (as I've explored locally). I'm down for just adding special-casing for `s-l` for now, with the cravats in mind that `s-e`, `s-l`, `s-f`, etc. still have the same problem. I'm also okay with following up with a new PR later, and just let me know how you'd like to proceed.

71 replied:
Weirdly enough, both my branch and yours have the exact same behavior on my end with your `h`/`l` example. I don't really get it, but let's go with the solution that works in two places rather than one :)

Much better when it comes to mouse handling!

It does sound like a lot of work. I'd like a temporary solution for `s-l` in the meantime, yes. I also created #53 in the meantime. Please add a comment around workaround / known bugs pointing to #53, and we'll tend to them later.

Once again, thank you for your work! I'll look at the changes and report on any changes that may be needed.

71 commented on src/commands/index.ts:
```diff
@@ -157,9 +157,15 @@ export class CommandDescriptor<Input extends InputKind = InputKind> {
+      // Force selection to be non-empty when switching to normal. This is only
+      // necessary because we do not restore selections yet.
+      // TODO: Remove this once https://github.com/71/dance/issues/31 is fixed.
+      state.normalizeSelections(editor)
```
This should be in [`Extension.setEditorMode`](https://github.com/71/dance/blob/747fb49e99c93ad8055e39471aff34be8abe70e2/src/extension.ts#L96), otherwise this doesn't happen when the extension is started, for instance.

yuchenshi replied:
Done.

71 commented on src/extension.ts:
```diff
@@ -255,6 +255,28 @@ export class Extension implements vscode.Disposable {
+  /**
+   * Make all selections in the editor non-empty by selecting at least one character.
+   */
+  normalizeSelections(editor: vscode.TextEditor) {
```
Since all callers will be in the same class, this can be made `private`.

yuchenshi replied:
Done.

71 commented on src/extension.ts:
```diff
@@ -255,6 +255,28 @@ export class Extension implements vscode.Disposable {
+   * Make all selections in the editor non-empty by selecting at least one character.
+   */
+  normalizeSelections(editor: vscode.TextEditor) {
+    if (this.modeMap.get(editor.document) !== Mode.Normal) { return; }
```
Code style (must also be applied below).

```suggestion
    if (this.modeMap.get(editor.document) !== Mode.Normal)
      return
```

yuchenshi replied:
Done.

71 commented on src/extension.ts:
```diff
@@ -255,6 +255,28 @@ export class Extension implements vscode.Disposable {
+   */
+  normalizeSelections(editor: vscode.TextEditor) {
+    if (this.modeMap.get(editor.document) !== Mode.Normal) { return; }
+    if (this.configuration.get('selections.allowEmpty')) { return; }
```
This call may be expensive, instead please use `observePreference` in the `constructor`.

yuchenshi replied:
Done.

71 commented on src/extension.ts:
```diff
@@ -255,6 +255,28 @@ export class Extension implements vscode.Disposable {
+  normalizeSelections(editor: vscode.TextEditor) {
+    if (this.modeMap.get(editor.document) !== Mode.Normal) { return; }
+    if (this.configuration.get('selections.allowEmpty')) { return; }
+    if (editor.selections.some(sel => sel.isEmpty)) {
```
Since this is called every time the selection changes, I'd like to avoid allocations as much as possible. Could you rewrite this so it only uses one loop and no `map` / `some`? For instance, create an array and populate it manually with a `for` loop, and use a boolean to know whether some changes were applied (i.e `selection.isEmpty` was true at some point). If that boolean is true at the end of the loop, set `editor.selections` to the array you built.

yuchenshi replied:
Done.

71 commented on src/extension.ts:
```diff
@@ -16,13 +16,17 @@ export class Extension implements vscode.Disposable {
   private configuration = vscode.workspace.getConfiguration(extensionName)
 
   enabled: boolean = false
+  private allowEmptySelections: boolean = true
```
Please add an empty line between `enabled` and these two fields.

71 commented on src/extension.ts:
```diff
@@ -238,6 +254,19 @@ export class Extension implements vscode.Disposable {
+              this.normalizeSelections(e.textEditor)
+              this.normalizeTimeoutToken = undefined
+            }, 200)
+          } else
```
Please wrap the `else` block in brackets.

71 commented on src/extension.ts:
```diff
@@ -260,6 +289,53 @@ export class Extension implements vscode.Disposable {
+            normalizedSelections.push(new vscode.Selection(active, active.translate(0, -1)))
+          }
+        }
+      } else if (normalizedSelections)
```
Please wrap the `else if` block in brackets.
```suggestion
      } else if (normalizedSelections !== undefined) {
```

71 commented on src/extension.ts:
```diff
@@ -260,6 +289,53 @@ export class Extension implements vscode.Disposable {
+
+    // Since this is called every time when selection changes, avoid allocations
+    // unless really needed and iterate manually without using helper functions.
+    let normalizedSelections
```
Please add a type annotation here (and explicitly initialize to `undefined`).
```suggestion
    let normalizedSelections: vscode.Selection[] | undefined = undefined
```

71 commented on src/extension.ts:
```diff
@@ -260,6 +289,53 @@ export class Extension implements vscode.Disposable {
+        normalizedSelections.push(selection)
+    }
+
+    if (normalizedSelections)
```
```suggestion
    if (normalizedSelections !== undefined)
```

71 commented on src/extension.ts:
```diff
@@ -260,6 +289,53 @@ export class Extension implements vscode.Disposable {
+
+    for (let i = 0; i < editor.selections.length; i++) {
+      const selection = editor.selections[i]
+      if (selection.isEmpty) {
```
Please add an empty line above.

yuchenshi commented:
I've implemented Kak-like behavior for one-character selections for `H`/`J`/`K`/`L` in e941a43e108987c4b1062450fd75e982af250ba0 and I believe this can be easily generalized to other commands when we want to fully fix #53. But for now, let's keep this PR less bloated.

I've also addressed the comments. PTAL.

71 commented on src/commands/move.ts:
```diff
@@ -138,15 +138,30 @@ function moveDown(editor: vscode.TextEditor, expand: boolean, mult: number) {
 
+/**
+ * Changes one character selections to be the direction specified. Return others unchanged.
+ * @param selection the original selection. Will be reversed if needed
```
Nit: Capitalize "Selection" and "Direction" below. Add empty line between the summary and the `@param`s. Add period at the end of the line.

yuchenshi replied:
Sorry but I cannot capitalize the `Selection` in `@param selection` -- it should be exactly the same as the parameter name based on how jsdoc works. But I've capitalized the descriptions for both parameters and made them full sentences because I think that's what you really meant.

71 commented on src/commands/move.ts:
```diff
@@ -138,15 +138,30 @@ function moveDown(editor: vscode.TextEditor, expand: boolean, mult: number) {
+  if (selection.isSingleLine && selection.active.character + direction === selection.anchor.character) {
+    // Treat one-character selection as non-directional when expanding.
+    // TODO(#53): Apply this to all extending commands in addition to HJKL.
+    return new vscode.Selection(selection.active, selection.anchor)
```
Works great on my end!

yuchenshi commented:
Rebased!

Sean1708 commented:
@yuchenshi When I set `dance.selections.AllowEmpty` to `false`, `Shift+H` and `Shift+L` stop working for me when in normal mode. `Shift+H` doesn't do _anything_ and `Shift+L` just moves the selection right without extending it. Is this something I'm doing wrong or is it an issue with the implementation? If the latter I'll open a ticket, but I didn't want to spam the issue tracker in case I was just doing something stupid...

Sean1708 commented:
I've played around with it a bit more and it looks like `w`, `e`, and `b` also behave weirdly; if the selection is only one character _and_ the last key you pressed was one of these then pressing one of them again won't do anything.

Delapouite commented:
> When I set dance.selections.AllowEmpty to false, Shift+H and Shift+L stop working for me when in normal mode. Shift+H doesn't do anything
I observed the same behavior

Sean1708 commented:
> I've played around with it a bit more and it looks like `w`, `e`, and `b` also behave weirdly

Actually it looks like this might be a more general issue, because it's happening even when this option is set to `true`... I'm starting to think this might be an issue on my end, but I have no idea how to debug it unfortunately :(

yuchenshi commented:
@Sean1708 @Delapouite I think what you've seen is similar to #69 and a few other recent issues. Would you mind chiming in there with more information like vscode version etc.?

Sean1708 commented:
No worries @71, happens to the best of us!

Thanks for fixing it so quickly!
